compiler: remove support for hot-code-reloading

zerbina opened the pull request:
## Summary
HCR was only tested by a single test and made the code-generator logic more complex. The feature is removed in order to reduce the amount of code in `cgen` and to make future changes in the code-generator easier.

The plan is to re-implement support for hot-code-reloading at a point in the future.

## Details
- remove the HCR compiler support
- remove the HCR runtime
- remove the enabled test. The disabled `nimhcr_integration.nim` test is kept
- remove the `nonReloadable` and `executeOnReload` pragmas

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

If wanted, I can also remove the `nimchr_integration.nim` test. The idea behind keeping it was to already have some existing tests when re-implementing HCR support.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on lib/system.nim:
```diff
@@ -529,10 +529,7 @@ proc `..`*[T](b: sink T): HSlice[int, T]
-  {.pragma: hcrInline, inline.}
-else:
-  {.pragma: hcrInline.}
+{.pragma: hcrInline.}
```
Could we get rid of this too?

zerbina replied:
Sure. I only kept it in order to reduce the size of the change-set.

saem replied:
Makes sense, but more feels is more points. 😁🥳

saem commented:
Nice!

Outside of CI failure I think it's good to go.

saem commented on lib/system/cgprocs.nim:
```diff
@@ -13,8 +13,8 @@ type
-proc nimLoadLibrary(path: string): LibHandle {.compilerproc, hcrInline, nonReloadable.}
-proc nimUnloadLibrary(lib: LibHandle) {.compilerproc, hcrInline, nonReloadable.}
-proc nimGetProcAddr(lib: LibHandle, name: cstring): ProcAddr {.compilerproc, hcrInline, nonReloadable.}
+proc nimLoadLibrary(path: string): LibHandle {.compilerproc, nonReloadable.}
```
Personally, I would remove `nonreloadable`, as well.

I know it increases the size, but it more importantly highlights how invasive this whole feature was in the first place.

saem replied:
this is non-blocking, I'm good to merge it.

zerbina replied:
It can go too - the required changes are of an okay amount.

saem replied:
Sweet!

zerbina commented:
The `.nonReloadable` pragma is now also removed. I missed removing the `executeOnReload` pragma earlier, but that's gone now too.

bors r+

saem commented:
Yaaaaaasssssssssss!

zerbina committed:
compiler: remove support for hot-code-reloading

## Summary
HCR was only tested by a single test and made the code-generator
logic more complex. The feature is removed in order to reduce the
amount of code in `cgen` and to make future changes in the
code-generator easier.

The plan is to re-implement support for hot-code-reloading at a point
in the future.

## Details
- remove the HCR compiler support
- remove the HCR runtime
- remove the enabled test. The disabled `nimhcr_integration.nim` test is
  kept
- the `.nonReloadable` and `.executeOnReload` pragmas are removed

saem commented:
bors r+

zerbina commented:
bors r+
